docs/design-documents/hal/mpu_design.md

```diff
@@ -0,0 +1,187 @@
+# MPU Execute Never Ram Support
+
+# Table of contents
+
+1. [Mbed OS design document](#mbed-os-design-document).
+1. [Table of contents](#table-of-contents).
+    1. [Revision history](#revision-history).
+1. [Introduction](#introduction).
+    1. [Overview and background](#overview-and-background).
+    1. [Requirements and assumptions](#requirements-and-assumptions).
+1. [System architecture and high-level design](#system-architecture-and-high-level-design).
+    1. [System architecture and component interaction](#system-architecture-and-component-interaction).
+1. [Detailed design](#detailed-design).
+1. [Usage scenarios and examples](#usage-scenarios-and-examples).
+
+### Revision history
+
+1.0 - Initial revision - Russ Butler - 10/17/2018.
+
+# Introduction
+
+### Overview and background
+
+This MPU support is intended to increase the security of mbed-os by restricting when code can execute from ram. One common attack on embedded devices is remote code execution by modifying ram. This happens by overflowing a ram buffer and then executing instructions from that buffer. By disabling execution from RAM mbed-os is more resistant against this form of attack.
+
+### Requirements and assumptions
+
+This API is limited to just enabling/disabling execution from RAM. It does not include runtime reconfiguration of the MPU regions.
+- Only standard Arm MPU supported initially
+- Only ARMv6m and ARMv7m devices with an MPU supported initially
+- Single common memory map used
+
+# System architecture and high-level design
+
+The proposed change consists of two APIs, one for porting and one for general use, system integration and initial porting work:
+
+ - HAL API so partners can port custom MPUs
+ - Platform API so applications and libraries can enable or disable the execution from RAM
+ - System integration so the execution from RAM is disabled by default
+ - Initial porting for the Arm MPU
+
+The goal of this is to increase security for a large number of devices by disabling execution from ram it is not required.
```

Comment:
`disabling execution from ram it is not required.` what's not required?

Reply:
Good catch, this should be `disabling execution from ram when it is not required`